### ISSUE TITLE
Rewrite to support async/ASGI Django (Waiting for Django ORM to be async)

### ISSUE THREAD

**Issue by JonasKs** [snok/django-auth-adfs#138]
**This issue is waiting for the Django ORM to be async and pushed to master**

We all had our hopes up for an async ORM to be pushed with Django 3.2, but this seems unlikely at the moment. [Alpha has been released](https://www.djangoproject.com/weblog/2021/jan/19/django-32-alpha-1-released/) without any mention of async, and the ORM [design has not been finalized](https://forum.djangoproject.com/t/asynchronous-orm/5925/15). 

I've created this issue to document and discuss anything related to this.

**Comment by JonasKs** [comment]
This can now be started. 

I am no longer using Django myself, but I'll be happy to review if someone wants to implement it. 

Example async aware middleware for Django can be found in our other repo [django-guid](https://github.com/snok/django-guid)

**Comment by jacoobes** [comment]
@JonasKs, hey, what parts of the codebase would need to be rewritten?

**Comment by JonasKs** [comment]
Hey @jacoobes , yeah, `LoginRequiredMiddleware` must be rewritten, but so must all the functionality in `backend.py`.  

If you'd like to work on this, I recommend doing a very small proof of concept and open a PR, in order to get feedback before too much time is spent.

**Comment by jacoobes** [comment]
Let me handle it!

**Comment by jacoobes** [comment]
Hey, would you guys prefer a breaking change (switch to full async) and remove wsgi or i could add another backend, ie AsyncXYZ ?

**Comment by JonasKs** [comment]
I think we must support both?

I'm worried about the test suite, though. Might be even harder to maintain..

**Comment by tim-schilling** [comment]
The test suite doubles unfortunately or requires a signficant rewrite to avoid it. We dealt with this on the toolbar's async GSoC project.

**Comment by jacoobes** [comment]
Alright. In addition, the codebase uses urlib3, i assume i would need to switch to an async client, im going to use httpx, if thats fine

**Comment by JonasKs** [comment]
`httpx` is fine, yes. 

I'd like to know a bit more about the performance impact having a sync middleware has in an async stack, though. Asgiref handles this today. 
If it doesn't solve or improve anything, adding the complexity isn't worth it, imo.

**Comment by jacoobes** [comment]
Yeah good point. If you enable debug logging for django request, you’ll see
that this middleware is auto adapted. Docs says it can support both, but at
a performance penalty.

**Comment by JonasKs** [comment]
Gotcha! 

Well, this is all yours if you'd like to give it a shot.   
Please keep us posted on the progress, and ask us questions if they come up. ☺️

**Comment by jacoobes** [comment]
This could be substantially less complex than we anticipated:
https://github.com/django/django/pull/18036
